### DISCUSSION TITLE
Release Testing 2023.05.1 - RC2

### DISCUSSION THREAD

**Discussion by tedpatrick** [pyscript/pyscript#1498] [Announcements]
We are feature complete for a release of PyScript. We need your help testing 2023.05.1.

How to test 2023.05.1.RC2

```
<link rel="stylesheet" href="https://pyscript.net/snapshots/2023.05.1.RC2/pyscript.css"/>
<script defer src="https://pyscript.net/snapshots/2023.05.1.RC2/pyscript.js"></script>
```

Thank you for your ongoing support and contribution to PyScript.

2023.05.1.RC2 Example:

```
<html>
<head>
    <link rel="stylesheet" href="https://pyscript.net/snapshots/2023.05.1.RC2/pyscript.css" />
    <script defer src="https://pyscript.net/snapshots/2023.05.1.RC2/pyscript.js"></script>
</head>
<body>
    <py-config type="toml">
packages = [
    "pandas"
]
    </py-config>
    <py-script>
import js
import pandas as pd

js.console.log( "2023.05.1.RC2" )
js.document.title = "2023.05.1.RC2"
print( "2023.05.1.RC2" )
print( pd.show_versions() )
    </py-script>
</body>
</html>
```

**Comment by bugzpodder** [comment]
[this](https://github.com/pyscript/pyscript/issues/1476#issuecomment-1553578648) is still an issue

**Comment by sskagemo** [comment]
When I use the RC2, I get errors when referring to global variables, and also globals(), in a py-repl.

You can check the code [here](https://nordicsmartgovernment.github.io/statistikkpilot_ssb/test.html) (replace the REPL-content with globals())

It has almost nothing but the REPL so I *hope* it is not due to some stupid error of my own :-)

This is the full error from the Chrome dev-tools console:
```
transfer_handlers.ts:134 Uncaught (in promise) PythonError: pyodide.ffi.JsException: Error: This borrowed proxy was automatically destroyed. Try using create_proxy or create_once_callable.
The object was of type "dict" and had repr "{'__name__': '__main__', '__doc__': None, '__package__': None, '__loader__': <class '_frozen_importlib.BuiltinImporter'>, '__spec__': None, '__annotations__': {}, '__builtins__': <module 'builtins' (built-in)>, '_pyodide_core': <module '_pyodide_core' (built-in)>, 'js': [object Window], 'pyscript': <module 'pyscript' from '/home/pyodide/pyscript/__init__.py'>, 'Element': <class 'pyscript._html.Element'>, 'display': <function display at 0x12fa858>, 'HTML': <class 'pyscript._html.HTML'>, 'write_to_page': <function write_to_page at 0x113b638>}"

    at new_error (pyodide.asm.js:9:14992)
    at pyodide.asm.wasm:0x152d67
    at pyodide.asm.wasm:0x152e6c
    at Module._pythonexc2js (pyodide.asm.js:9:654999)
    at Module.callPyObjectKwargs (pyodide.asm.js:9:75811)
    at Proxy.callKwargs (pyodide.asm.js:9:89719)
    at innerMessageHandler (async_task.ts:84:31)
    at async_task.ts:154:21
    at Generator.next (<anonymous>)
    at pyconfig.ts:266:1
    at new Promise (<anonymous>)
    at __async (pyconfig.ts:266:1)
    at callback (async_task.ts:130:76)
    at FakeMessagePort.postMessage (fake_message_channel.ts:30:7)
    at requestResponseMessage (request_response.ts:31:6)
    at SynclinkTask.<anonymous> (task.ts:189:24)
    at Generator.next (<anonymous>)
    at pyconfig.ts:266:1
    at new Promise (<anonymous>)
    at __async (pyconfig.ts:266:1)
    at SynclinkTask.do_async (task.ts:188:32)
    at SynclinkTask.schedule_async (task.ts:81:10)
    at SynclinkTask.<anonymous> (task.ts:100:10)
    at Generator.next (<anonymous>)
    at pyconfig.ts:266:1
    at new Promise (<anonymous>)
    at __async (pyconfig.ts:266:1)
    at SynclinkTask.then (task.ts:99:17)
```

Please let me know if you need this report somewhere else.

And I just have to say that I really love PyScript :-)

**Reply by sskagemo** [reply]
Btw, I tried help(globals) in the REPL in the page I linked to above and I got the help-text, but it is repeated (three times) and comes out both in the main page and in a console (se screenshot).

Don't know if it is related to the above problem, though.

![image](https://github.com/pyscript/pyscript/assets/5012227/7db49723-d51f-4328-bbca-af51159a4509)

**Reply by JeffersGlass** [reply]
I don't think this second issue is related to `help` specifically, but is related to lifetime management of PyProxies somehow. When I run `globals()` in a REPL I see the same error as you; if I then run anything that prints to stdout (e.g. `print("foo")`) it displays to the page multiple times.
![image](https://github.com/pyscript/pyscript/assets/1931111/9e56f60b-92e3-4ae1-86c6-8e44a3365358)

Also, it's not just `globals()` - it seems to be anything that causes the REPL to return an object that becomes a PyProxy. Things I've tried that get the same error:

```python
globals()

dir()

[1,2,3]

class C:
   ...
C()

# deliberately mis-formatted - shows same synclink error instead of Python error
"Hello 
```

But these work fine
```python
42

"Hello World"

from js import document
document # [object HTMLDocument]

1/0 # ZeroDivisionError: division by zero
```

**Reply by JeffersGlass** [reply]
Doing a git bisect, this issue appears starting in #1318 .

**Reply by WebReflection** [reply]
I don't know if this is related, but I see a lot of non-cleared timers [in this change](https://github.com/pyscript/pyscript/commit/e9122bca9da119c7b6f4e4c4816b0adefee7ea27) ... most notably, if we use `setTimeout` for anything but we never clear it, we might have issues at distance.

Can anyone observe any delay between prints?

**Reply by WebReflection** [reply]
btw, I could reproduce the issue before but now I cannot reproduce it anymore ... I wonder if there was some issue at the CDN level

**Reply by WebReflection** [reply]
I cannot reproduce the (duplication) issue even via the original test page https://nordicsmartgovernment.github.io/statistikkpilot_ssb/test.html

**Reply by sskagemo** [reply]
Strange. I tested it and experienced the same myself. But then I got the error again. It seems that you have to get the first errror (calling globals() for instance) first, and then that will result in multiple copies when you run print("foo"). It seems that the more times I run globals() to get the first issue, the more copies I get when I run print("foo").

**Reply by WebReflection** [reply]
Yes, I can confirm to reproduce the duplication the initial error must be triggered ... it looks to me we're doing double work due dangling listeners but I couldn't find the culprit beside observing that the error is not handled correctly on PyScript side, it's thrown right in pyodide in the midlde of some complicated "dance".

**Reply by ntoll** [reply]
@hoodmane any thoughts on this?

**Reply by hoodmane** [reply]
I'll look into it.

**Reply by madhur-tandon** [reply]
@hoodmane a closer look here: https://37c8e61f-8cf6-4517-9652-2483b9db9016.pyscriptapps.com/ae4f15e8-c053-49b6-890b-8351fddbd122/latest/

Here, since we have error 2 times, the print will be 3 times, but in different cells.

**Reply by hoodmane** [reply]
Thanks @madhur-tandon!

**Reply by hoodmane** [reply]
Ok I am pretty sure I have a fix for this.

**Comment by JeffersGlass** [comment]
Do we have a concise explanation for [the need to await accesses to the Python runtime](https://github.com/pyscript/pyscript/discussions/1489#discussioncomment-6019586)? We don't have it in the documentation anywhere - and, selfishly, I'd like to understand it better so I can document it in a release post.

**Reply by madhur-tandon** [reply]
Pushing a fix for this soon @JeffersGlass

**Reply by madhur-tandon** [reply]
https://github.com/pyscript/pyscript/pull/1529 should fix it

**Reply by JeffersGlass** [reply]
Looks great! Thanks @madhur-tandon !

**Comment by tedpatrick** [comment]
Published RC3
https://github.com/pyscript/pyscript/discussions/1541

**Reply by sskagemo** [reply]
RC3 fixed my problem, as far as I can see -- thank you :-)

**Reply by hoodmane** [reply]
Thanks again for the report @sskagemo!
